### ISSUE TITLE
Maybe crashes from exwm

### ISSUE THREAD

**Issue by spiderbit** [ch11ng/exwm#209]
Hmm I am not shure if that is a bug report or just a way to figure out if a bug exists in exwm that causes crashes.

So I have 3 sorts or crashes:

1. browser windows somethimes crash often when I am in input boxes as example youtube comments
That could be browserspecific, but maybe It has also to do with blocking in emacs or something.
2. I maybe land cant really say exactly how often 5-10 times maybe a weak on the console or Systemd restarts the login X manager. I had this in fedora and in nixos now.
As far as I remember I dont have that issue on my chromebook. So it could be a hardware problem, not shure or its related to gnus, cause I dont use gnus on the chromebook.
3. somethimes my PC just restarts, here I am pretty shure its not exwms fault, maybe has to do with the docking station or overheating or acpi errors something like that.

A. I just wunder if somebody Else has here and there crashing exwm on them or am I alone on that? I wonder if hardware could cause that (I dont belive so I mean I run a memtest already, all fine) or maybe gnus.

B. If gnus could be the problem, would it maybe make sense to have a seperate emacs inside a X window, but with other keybindings, that dont conflict with exwm, that would be not really nice to have but if it saves me from crashes maybe worth.

**Comment by 11111000000** [comment]
Thinkpad T420, Thinkpad X220T. Very clean Arch. No gnome & other crap. Pure Xorg + EXWM. Emacs config something like in my repo + exwm. Everyday tasks works like charm (surprisingly), but sometimes `google-chrome-stable` hangs (exwm hangs too) and we can't start it again, even after  Xorg restart.  `google-chrome-stable` can start again only after full system reboot. Now I cant say, why. Today I do M-<F2>, then `killall -CONT emacs` - it seems like this helps to unfreze exwm in some cases. 
Maybe any suggestions to setup "magic-key" (though `xbindkeys`) to gracefuly restart crashed `exwm`?

**Comment by 11111000000** [comment]
Some Chrome hangs may related to 3d-acceleration. Try that:

/etc/X11/xorg.conf.d/10-intel.conf
```
Section "Device"
	Identifier  "Intel Graphics"
	Driver      "intel"
	Option	    "DRI"	"false"
  Option      "TearFree"    "true"
  Option      "NoAccel" "True"
EndSection
```
/etc/modprobe.d/i915.conf
```
options i915 enable_rc6=0 enable_fbc=1 i915.enable_psr=0 semaphores=1
```

Before, I see flicker with `google-chrome` on `M-x` menu shown, now it works more smoothly.

**Comment by ch11ng** [comment]
@spiderbit,

I've not experienced any crash recently and I'm a bit surprised to see the report.  It at least has something to do with Emacs and perhaps EXWM just triggers a certain bug.  It would be helpful if you can save the error stream of Emacs.  As for Gnus, you can start it in a separate Emacs instance, make the instance fullscreen and put it in a dedicated workspace.  It may not be Gnus' fault though.

**Comment by ch11ng** [comment]
@11111000000,

Similar problems seem only have been reported for Google Chrome so maybe it's very specific to it.  You can try setting up a log file for Chrome and hopefully it will tell something.

You need a session manager to saving/restoring opened X windows.  It's currently not implemented in EXWM but I think it does not solve this problem.

I've seen someone making Emacs a systemd daemon before.  That seems not difficult.  Also it doesn't solve the problem.

**Comment by spiderbit** [comment]
I mean I dont use google chrome but qutebrowser but likely qutebrowser uses webkit, so it could be related.

what are this kind of "hangs" application crashes? X crashes? System crashes, what are we talking about?

But its interesting that you are talking about intel when I think did not mention what hardware I have and yes indead its a intel notebook. that said my chromebook is intel too, so should it not happen on that machine too?

The 2 things that are majorly different are 1. gnus, which creates some blocking here and there 2. hardware. But again I dont see hardware issues crashing X if its not memory problems, I would suspect then systemhangs or sudden reboots, which again happen here too from time to time.

https://wiki.archlinux.org/index.php/Intel_graphics#X_freeze.2Fcrash_with_intel_driver

so ok that is a thing aperently? 
is that: enable_rc6=0 a alternative to  intel_idle.max_cstate=1 what they mention? rc6 is set to 1 at the moment.

I aperently trust X and kernel developers to much when they dont choose save default values or intel breaks stuff left and right on a 100 year old hardware (2012).

fbc should be deactivated by default since kernel 4.6 since a phoronix article.

well aperently not:
/sys/module/i915/parameters/enable_fbc:-1

interesting, didnt suspect such problems. its always that mantra that intel has good (free) linuxdrivers, you think you just cant have problems accept features/speed with free drivers, on intels side apperently thats not so true.

thanks for giving me that hint. Maybe its good to keep this bug open so I could confirm that I get rid of the X restarts at least, with that method, if you wanna close it, its also ok, then I would reopen it if thats not the case.

I would then maybe use a few days something else then exwm (even that would be torture) and see if I have there similar problems or not or test a seperate gnus window.

**Comment by spiderbit** [comment]
Well I have a Thinkpad X220, too so same hardware.

I wonder if your:

```
  Option      "DRI"   "false"  
  Option      "TearFree"    "true"
```

are needed cause TearFree seems to be only needed if SNA accelleration makes problems, but you have NOACCEL activated so you are not using SNA.

Also if you have deactivated ACCEL you dont have to seperatly deactivate DRI as far as I understand at least that is what the log tells me:

```
[     4.026] (WW) intel(0): loading DRI2 whilst acceleration is disabled.
[     4.026] (II) intel(0): [DRI2] Setup complete
[     4.026] (II) intel(0): [DRI2]   DRI driver: i965
[     4.026] (II) intel(0): [DRI2]   VDPAU driver: va_gl
[     4.026] (II) intel(0): direct rendering: DRI2 enabled
```

and I only have noaccel set to true.

Also I dont have enable_psr = 0 in the module options cause at least in nixos that seems to be default, and I think thats upstream default too, at least the archwiki does not mention it.

1. Day no system crash so far, but browser crashed, so that seems at least not related to that (I have another bugreport about that browser open somewhere else) But if that "solves" my Xorg crashes that would be great. :)

**Comment by spiderbit** [comment]
Well Update 10 hours later again a X crash, with Noaccel in xorg conf and following modprobe line:

`Option "NoAccel" "True"`

`options i915 enable_rc6=1 enable_fbc=1 semaphores=1`

So I think exwm bug is more likely, I think I had such crashes too when I used conkeror a xulrunner based browser, like now with qutebrowser.

So either its exwm or gnus (which kind of would be exwms fault too :) )

**Comment by ch11ng** [comment]
@spiderbit,

So it's Emacs not Xorg that crashes.  I think you need to check the error stream of Emacs:

```shell
emacs >> ~/exwm.log 2>&1
```

**Comment by spiderbit** [comment]
hmm I use the nixos magic that packaged up exwm as distropackage, no manual .xinitrc shit anymore... so with lightdm starting it.

But I am now interested if and which of the 2 emacs instances crashes now. I seperated like you suggested gnus into a seperate workspace/emacs instance.

should have another crash in next few days if its not gnus fault... then I try your next suggestion :)

**Comment by spiderbit** [comment]
I am not 100% shure, I had to restart PC  / X unrelated to this a few times so it could only be random luck, but I think using a seperate emacs for gnus helped.

I cant even remember browser crashes not to mention X crashes or system crashes.

I will not make a final call yet, but I think that solved my problems.

**Comment by medranocalvo** [comment]
I guess this is solved.  Please reopen if that's not the case.
